frontend/src/components/matching/MatchingLobby.tsx

```diff
@@ -44,143 +39,175 @@ export default function MatchingLobby({
-    setStage(next)
+  const initialSuccessState: MatchingSuccessState = {
+    userReady: false,
+    partnerReady: false,
+    partnerLeft: false,
+    owner: false,
+    partner: {
+      id: "",
+      name: "",
+      image: undefined
+    }
   }
 
-  // Trigger matching process on modal open
-  React.useEffect(() => {
-    if (isOpen) {
-      setStage(MATCHING_STAGE.MATCHING)
-    } else {
-      setStage(MATCHING_STAGE.INITIAL)
-    }
-  }, [isOpen])
+  const [stage, setStage] = React.useState(MATCHING_STAGE.INITIAL);
+  const [successState, setSuccessState] = React.useState(initialSuccessState);
 
-  // Response to stage events
-  React.useEffect(() => {
-    switch (stage) {
-      case MATCHING_STAGE.MATCHING:
-        onMatchingStage();
-        break;
-      default:
-        break;
-    }
-  }, [stage])
+  const [socket, setSocket] = React.useState<Socket>();
+  const { user } = useAuthContext();
 
-  // Handle view switching
-  const renderView = (stage: MATCHING_STAGE) => {
-    switch (stage) {
-      case MATCHING_STAGE.MATCHING:
-        return initialView;
-      case MATCHING_STAGE.SUCCESS:
-        return successView;
-      case MATCHING_STAGE.FAIL:
-        return failureView;
-      default:
-        return errorView;
+  const logger = getLogger('matching');
+
+  /////////////////////////////////////////////
+  // Stage fired events
+  /////////////////////////////////////////////
+  const initializeSocket = async () => {
+    try {
+      const config = await getMatchingSocket();
+      const socket = io(config.endpoint, { path: config.path });
+
+      // Register server events
+      socket.on('connect', () => setStage(MATCHING_STAGE.MATCHING));
+      socket.on("disconnect", () => setStage(MATCHING_STAGE.ERROR));
+      socket.on("connect_error", () => setStage(MATCHING_STAGE.ERROR));
+      socket.on("matched", handleMatched);
+      socket.on("no_match", handleNoMatch);
+      socket.on("room_closed", handleRoomClosed);
+      socket.on("peer_ready_change", handlePartnerReadyChange);
+
+      setSocket(socket);
+    } catch (error) {
+      logger.error(error);
+      setStage(MATCHING_STAGE.ERROR)
     }
   }
 
-  // Handles matching event
   const onMatchingStage = () => {
     try {
-      console.log("matching process triggered");
-
-      // Request to join matching queue
-      setTimeout(() => {
-        setStage(MATCHING_STAGE.SUCCESS)
-      }, 60 * 1000)
+      setSuccessState(initialSuccessState);
+      socket?.emit("request_match", {
+        user: {
+          id: user.id,
+          name: user.name,
+          image: user.image
+        },
+        preferences: options
+      });
     } catch (error) {
-      setStage(MATCHING_STAGE.ERROR)
+      logger.error(error);
+      setStage(MATCHING_STAGE.ERROR);
     }
   }
 
-  // Handles retry action
-  const handleRetry = () => {
-    setStage(MATCHING_STAGE.MATCHING);
+  /////////////////////////////////////////////
+  // Server fired events
```

Comment:
For all the event handlers below, do you consider to move them under `helpers`? You can put them under `api_wrapper` for matching service? What do you think?